### PR TITLE
Add NOI Cards

### PR DESCRIPTION
**Pull request by dhaselhan**

### CODE REVIEW

File: services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  ) {}
+
+  async create(data: CreateNoticeOfIntentDto, board: Board) {
+    const existingNoticeOfIntent = await this.repository.findOne({
+      where: {
+        fileNumber: data.fileNumber,
+      },
+    });
+
+    //TODO: Also Check Covenants?
```
Comment (sandratoh) [inline services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts:53]: Is this for future ticket?

File: services/apps/alcs/src/alcs/board/board.controller.spec.ts [hunk 2]
```diff
@@ -28,15 +29,17 @@ describe('BoardController', () => {
...
-    cardService = createMock<CardService>();
-    covenantService = createMock<CovenantService>();
+    boardService = createMock();
+    appService = createMock();
+    appReconsiderationService = createMock();
+    modificationService = createMock();
+    planningReviewService = createMock();
+    cardService = createMock();
+    covenantService = createMock();
+    noticeOfIntentService = createMock();
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply services/apps/alcs/src/alcs/board/board.controller.spec.ts:42]: The type is already declared on the let clause above, so having it here is redundant. We stopped adding these a while back but this file hasn't been touched for a while.

File: services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts (hunk 1 above)
Comment (dhaselhan) [reply services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts:53]: No, this is for me to ask Mekthi what we think we should when he is back, I'm not quite sure on the best approach.

File: alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      throw e;
+    }
+  }
+
+  async fetchByCardUuid(id: string) {
+    try {
+      return await firstValueFrom(this.http.get<NoticeOfIntentDto>(`${this.url}/card/${id}`));
+    } catch (e) {
+      console.error(e);
+      this.toastService.showErrorToast('Failed to fetch covenant');
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.ts]: Done!

**Comment by sandratoh** [comment]
LGTM but shall wait for @mhuseinov's  review 👀

File: alcs-frontend/src/app/features/board/board.component.ts [hunk 4]
```diff
@@ -222,12 +238,14 @@ export class BoardComponent implements OnInit, OnDestroy {
       const sorted = [];
       sorted.push(
         // high priority
+        ...mappedNoticeOfIntents.filter((a) => a.highPriority).sort((a, b) => b.activeDays! - a.activeDays!),
```
Comment (mhuseinov) [inline alcs-frontend/src/app/features/board/board.component.ts:241]: I thought that NOIs are always at the top no matter the priority and then applications and the rest type of cards

File: alcs-frontend/src/app/features/board/dialogs/notice-of-intent/notice-of-intent-dialog.component.scss [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  max-height: 250px;
+}
+
+.board-menu-item {
+  display: grid;
+  grid-template-columns: 22px 1fr 22px;
+  width: 100%;
+}
+
+::ng-deep .mat-mdc-menu-item .mdc-list-item__primary-text {
```
Comment (mhuseinov) [inline alcs-frontend/src/app/features/board/dialogs/notice-of-intent/notice-of-intent-dialog.component.scss:59]: no :host ?

File: alcs-frontend/src/app/features/home/assigned/assigned.component.ts [hunk 6]
```diff
@@ -101,7 +101,7 @@ export class AssignedComponent implements OnInit {
     this.sortedFiles = sorted;
   }
 
-  private mapCovenant(c: CovenantDto): AssignedToMeFile {
+  private mapCovenant(c: NoticeOfIntentDto): AssignedToMeFile {
```
Comment (mhuseinov) [inline alcs-frontend/src/app/features/home/assigned/assigned.component.ts:104]: is this correct that mapCovenant accepts NoticeOfIntentDto?

File: alcs-frontend/src/app/services/home/home.service.ts [hunk 7]
```diff
@@ -22,7 +22,7 @@ export class HomeService {
         reconsiderations: ApplicationReconsiderationDto[];
         planningReviews: PlanningReviewDto[];
         modifications: ApplicationModificationDto[];
-        covenants: CovenantDto[];
+        covenants: NoticeOfIntentDto[];
```
Comment (mhuseinov) [inline alcs-frontend/src/app/services/home/home.service.ts:25]: looks a bit confusing to have covenants of type NoticeOfIntentDto[]

File: alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  it('should show an error toast message if create fails', async () => {
+    httpClient.post.mockReturnValue(
+      throwError(() => {
+        new Error('');
+      })
+    );
+
+    try {
+      await service.create({
```
Comment (mhuseinov) [inline alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.spec.ts:64]: is "expect(await service.create()).rejects.toMatchObject..." is not working?

File: alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+  constructor(private http: HttpClient, private toastService: ToastService) {}
+
+  async create(createDto: CreateNoticeOfIntentDto) {
+    try {
+      return await firstValueFrom(this.http.post<NoticeOfIntentDto>(`${this.url}`, createDto));
+    } catch (e) {
+      console.error(e);
+      if (e instanceof HttpErrorResponse && e.status === 400) {
+        this.toastService.showErrorToast(`Covenant or Application with File ID ${createDto.fileNumber} already exists`);
```
Comment (mhuseinov) [inline alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.ts:22]: totally a nit pick but, maybe something like this "Covenant/Application/NOI with File ID ${createDto.fileNumber} already exists"

File: services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts (hunk 1 above)
Comment (mhuseinov) [reply services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts:53]: discussed with Daniel the approach, will be a service that will be used for all entities that require the file number to be unique.

File: services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.entity.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  @Column({ unique: true })
+  fileNumber: string;
+
+  @Column()
+  applicant: string;
+
+  @Column({ type: 'uuid' })
+  cardUuid: string;
```
Comment (mhuseinov) [inline services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.entity.ts:24]: I wonder if card needs to be nullable the same way as in application. So when we move things from OATS we will not be creating cards for all the data?

File: alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.ts (hunk 9 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.ts:22]: Done!

File: alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.spec.ts (hunk 8 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/services/notice-of-intent/notice-of-intent.service.spec.ts:64]: Cleaned up

File: alcs-frontend/src/app/services/home/home.service.ts (hunk 7 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/services/home/home.service.ts:25]: Fixed

File: alcs-frontend/src/app/features/home/assigned/assigned.component.ts (hunk 6 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/features/home/assigned/assigned.component.ts:104]: Fixed

File: alcs-frontend/src/app/features/board/dialogs/notice-of-intent/notice-of-intent-dialog.component.scss (hunk 5 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/features/board/dialogs/notice-of-intent/notice-of-intent-dialog.component.scss:59]: Fixed

File: alcs-frontend/src/app/features/board/board.component.ts (hunk 4 above)
Comment (dhaselhan) [reply alcs-frontend/src/app/features/board/board.component.ts:241]: Not 100% sure, but with Ansley away we can fix this later if its wrong.

File: services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts (hunk 1 above)
Comment (dhaselhan) [reply services/apps/alcs/src/alcs/notice-of-intent/notice-of-intent.service.ts:53]: Going to do this in a future MR.

**Review by mhuseinov — APPROVED** [review]
🚀

**Pull request merged**